### DISCUSSION TITLE
Issue with shutil.rmtree() (Python 3.13) in a Fedora build environment and PyFakeFS involved

### DISCUSSION THREAD

**Discussion by buhtz** [jmcgeheeiv/pyfakefs#1084] [Q&A]
Hello,
I would like to request assistance with analysing a complex issue.

Maybe you have an idea or details what is going on. Currently it is not clear if PyFakeFS is a factor here. We also testing if the problem ocurse without it.

# Summary
A unit test is using PyFakeFS and `tempfile.TemporaryDirectory()` in a `with` context.
When leaving the context `shutil.rmtree()` is used to clean up the temp-dir.
In very specific environment this crashes. The environment is:
- Python 3.13 (crashing code was introduced within [PR cpython#119808](https://github.com/python/cpython/pull/119808))
- Fedora 41 build environment, which is using `mock` (a tool not a Python package) which create a `chroot` environment
- Maybe PyFakeFS: We will test if this problem ocure without PyFakeFS.

# More details
The initial report was https://github.com/bit-team/backintime/issues/1911

The related unit test

https://github.com/bit-team/backintime/blob/c1d042ab67b9e117ac53e944518a0f4292fa075b/common/test/test_uniquenessset.py#L45

The CPython shutil code causing that crash

https://github.com/python/cpython/blob/75401febc91a449cc4f4391d663e9a96ce91bb6c/Lib/shutil.py#L695

Ongoing discussion about the issue also on [python mailing list](https://mail.python.org/pipermail/python-list/2024-October/912881.html) and the [mock project](https://github.com/rpm-software-management/mock/discussions/1491).

**Comment by mrbean-bremen** [comment]
I checked the [log file](https://github.com/user-attachments/files/17516354/build_local_pychanged.log) with the failing tests mentioned in the linked issue and noticed that it runs under Python 3.13, but uses pyfakefs 5.2.4, which did not have support for Python 3.12 and 3.13 (it was published before  the release of Python 3.12).

Can you check if the problem also happens with older Python versions and /or if it occurs with a current pyfakefs version (3.7.1)? Preliminary support for 3.13 has only been introduced in 5.6, and official support with 5.7 (with the release of Python 3.13.0).

**Reply by buhtz** [reply]
Hell, yeah! 🚀 
Based on your comment I downgraded pyfakefs to version 5.2.4 in my own TravisCI config and run Python 3.9 up to Python 3.13 on an Ubuntu 22 VM.

The same test from the original reported issue fails with Python 3.13 and 3.12.
Nice.

So to my understanding the dependency rule should be if Python 3.12 or newer PyFakeFS should be minimal version 5.6 ?

Fedora 41 (where the report comes from) is still at PyFakeFS 5.2.4. I will report back to them.

**Reply by mrbean-bremen** [reply]
Actually minimal version for Python 3.13 should be 5.7 - the Python 3.13 qualifier was only added to this version. 5.6 only had some preliminary support before the final 3.13 was out.
Python 3.12 was supported since pyfakefs 5.3 (as can be seen in the [release notes](https://github.com/pytest-dev/pyfakefs/blob/main/CHANGES.md)).
